src/main/kotlin/io/provenance/p8e/plugin/Bootstrapper.kt

```diff
@@ -62,64 +154,298 @@ internal class Bootstrapper(
+                            .setResultType(DefinitionType.DEFINITION_TYPE_PROPOSED)
+                            .addResponsiblePartiesValue(functionSpec.invokerPartyValue)
                             .build()
+
+                        recordSpecifications.add(recordSpec)
+                    }
+
+                    Pair(contractSpecification, recordSpecifications)
+                }.also { specs ->
+                    project.logger.info("Adding ${specs.size} contract specification(s) to batch")
+
+                    specs.map { (contractSpecification, recordSpecifications) ->
+                        tx.add(
+                            MsgWriteContractSpecificationRequest.newBuilder()
+                                .addSigners(pbAddress)
+                                .setSpecification(contractSpecification)
+                                .build()
+                        )
+
+                        tx.addAll(recordSpecifications.map { spec ->
+                            val contractSpecAddress = MetadataAddress.fromBytes(spec.specificationId.toByteArray())
+
+                            MsgWriteRecordSpecificationRequest.newBuilder()
+                                .addSigners(pbAddress)
+                                .setSpecification(spec)
+                                .setContractSpecUuid(contractSpecAddress.getPrimaryUuid().toString())
+                                .build()
+                        })
+                    }
+                }
+
+                // add all new scope spec to contract spec mappings
+                newContractSpecifications.forEach { (spec, _) ->
+                    val address = MetadataAddress.fromBytes(spec.specificationId.toByteArray())
+
+                    contractSpecificationUuidToScopeNames.getValue(address.getPrimaryUuid()).forEach { scopeSpecName ->
+                        val uuid = scopeSpecificationNameToUuid.getValue(scopeSpecName)
+                        val scopeSpecificationAddress = MetadataAddress.forScopeSpecification(uuid)
+
+                        tx.add(
+                            MsgAddContractSpecToScopeSpecRequest.newBuilder()
+                                .setScopeSpecificationId(ByteString.copyFrom(scopeSpecificationAddress.bytes))
+                                .setContractSpecificationId(spec.specificationId)
+                                .addSigners(pbAddress)
+                                .build()
+                        )
                     }
+                }
+
+                // TODO handle adding ContractSpecToScopeSpec's for existing ContractSpecifications that add a new ScopeSpecification
+
+                if (tx.isNotEmpty()) {
+                    val serviceClient = ServiceGrpc.newBlockingStub(provenanceChannel)
+                    val authClient = cosmos.auth.v1beta1.QueryGrpc.newBlockingStub(provenanceChannel)
+
+                    tx.chunked(25).forEach { batch ->
```

Comment:
are transactions always processed in order according to the sequence number? I.e. if the first batch contained a bunch of contractspecs, and the second batch had recordspecs referencing those contractspecs, would the contractspecs be created already?

Reply:
oh wait... we are blocking, nevermind